### PR TITLE
delegate to custom completer when available

### PR DESCRIPTION
**Pull request by kevinushey**
This PR augments RStudio's completion system such that it delegates to the `custom.completer` R function when available.

### CODE REVIEW

**Comment by jmcphers** [comment]
Couple of quick questions:

a) If I'm reading this correctly then a custom R completion function, if enabled, will completely disable our own completion system and use the R completer instead. Does that sound right?

b) What mechanism sets/unsets `utils::rc.getOption("custom.completer")`? (Just want to be sure we're not going to wind up with users accidentally getting stuck with a custom completer...)

File: src/cpp/session/modules/SessionRCompletions.R [hunk 1]
```diff
@@ -1767,6 +1768,12 @@ assign(x = ".rs.acCompletionTypes",
    # parameterized R Markdown documents
    if (.rs.injectKnitrParamsObject(documentId))
       on.exit(.rs.removeKnitrParamsObject(), add = TRUE)
+
+   # If custom completions have been set through
+   # 'rc.options("custom.completions")',
+   # then use the internal R completions instead.
+   if (.rs.isCustomCompletionsEnabled())
+      return(.rs.getCustomRCompletions(line))
```
Comment (jmcphers) [inline src/cpp/session/modules/SessionRCompletions.R]: Should we fall back on our on completion system if the custom completer generates an error?

**Comment by kevinushey** [comment]
> a) If I'm reading this correctly then a custom R completion function, if enabled, will completely disable our own completion system and use the R completer instead. Does that sound right?

That's correct (and intentional, for this particular kind of completer).

> b) What mechanism sets/unsets utils::rc.getOption("custom.completer")? (Just want to be sure we're not going to wind up with users accidentally getting stuck with a custom completer...)

Typically, an R function will set it and unset it later with e.g. an `on.exit()` handler. For example, the v8 package does this:

https://github.com/jeroen/V8/blob/e6e60a7158c6746b992d69ab14bafd8f98864731/R/V8.R#L192-L196

to register a custom completer while the v8 REPL is running.

File: src/cpp/session/modules/SessionRCompletions.R (hunk 1 above)
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R]: Good idea!

**Comment by kevinushey** [comment]
As I understand it, the `custom.completer` option was designed to be used as a replacement for whatever the default completion system is, not work in tandem with it. I think if we want to support extensions to our own completion system it would have to exist separate from this specific option / capability of the R completion system.

The interface I have sitting around in my head would be to have R packages include something like:

```
.onLoad <- function(libname, pkgname) {
    rstudioapi::registerCompleter("fnName", function(context) { ... })
}
```

and our completion system could look up and use these functions when retrieving completion results for e.g. a function called `fnName`.

**Comment by jjallaire** [comment]
Okay, I agree this should be a wholesale replacement then and we can design
a separate interface for something that works in tandem

On Thu, Apr 6, 2017 at 9:28 PM, Kevin Ushey <notifications@github.com>
wrote:

As I understand it, the custom.completer option was designed to be used as
a replacement for whatever the default completion system is, not work in
tandem with it. I think if we want to support extensions to our own
completion system it would have to exist separate from this specific option
/ capability of the R completion system.

The interface I have sitting around in my head would be to have R packages
include something like:

.onLoad <- function(libname, pkgname) {
    rstudioapi::registerCompleter("fnName", function(context) { ... })
}

and our completion system could look up and use these functions when
retrieving completion results for e.g. a function called fnName.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/rstudio/rstudio/pull/1114#issuecomment-292390713>, or mute
the thread

File: src/cpp/session/modules/SessionRCompletions.R [hunk 2]
```diff
@@ -1767,6 +1768,20 @@ assign(x = ".rs.acCompletionTypes",
...
+   # 'rc.options("custom.completions")',
+   # then use the internal R completions instead.
+   if (.rs.isCustomCompletionsEnabled()) {
+      
+      completions <- tryCatch(
+         .rs.getCustomRCompletions(line),
+         error = identity
+      )
+      
+      if (!inherits(completions, "error"))
```
Comment (jmcphers) [inline src/cpp/session/modules/SessionRCompletions.R]: Should this error be logged somewhere? I'm thinking of the case where someone is expecting custom completions but not getting them, and doesn't know whether (a) they aren't being registered correctly, or (b) they are registered but triggering errors when run.

**Pull request merged**